src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValues.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.vectorvalues;
+
+import lombok.ToString;
+import org.apache.lucene.codecs.KnnFieldVectorsWriter;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.ByteVectorValues;
+import org.apache.lucene.index.FloatVectorValues;
+
+import java.io.IOException;
+
+/**
+ * An abstract class to iterate over KNNVectors, as KNNVectors are stored as different representation like
+ * {@link BinaryDocValues}, {@link FloatVectorValues}, {@link ByteVectorValues}, {@link KnnFieldVectorsWriter} etc.
+ * @param <T>
+ */
+@ToString
+public abstract class KNNVectorValues<T> {
+
+    protected final KNNVectorValuesIterator vectorValuesIterator;
+    protected int dimension;
+
+    protected KNNVectorValues(final KNNVectorValuesIterator vectorValuesIterator) {
+        this.vectorValuesIterator = vectorValuesIterator;
+    }
+
+    /**
+     * Return a vector reference. If you are adding this address in a List/Map ensure that you are copying the vector first.
+     * This is to ensure that we keep the heap and latency in check by reducing the copies of vectors.
+     *
+     * @return T an array
+     * @throws IOException if we are not able to get the vector
+     */
+    public abstract T getVector() throws IOException;
+
+    /**
+     * Dimension of vector is returned. Do call getVector function first before calling this function otherwise you will get 0 value.
```

Comment:
Would be nice if we can avoid this restriction. If getVector function is not called, we should call it before returning dimension.

Reply:
This restriction mainly comes from BinaryDocValues. As, before reading the BDV we don't know what the vector is and hence we don't know what is the dimension. For FVV, and ByteVV this is not a problem because the dimension gets serialized with VV.

If we want to fix this, we have to read a doc and put some complexity in the nextDoc functions. if we are okay with complexity on the nextDoc for BDV I can make the change. 

Another option is have a cloneable interface on BDV which we Lucene doesn't provide. So only option is have make the nextdoc function complex. What your thought on this?

@heemin32

Reply:
Would it be okay for caller to determine dimension using the returned vector? Does dimension have to be inside this class?

Reply:
> Would it be okay for caller to determine dimension using the returned vector? Does dimension have to be inside this class?

So for binary vector they need to identify the type of vector and then use the dimension. Having it like this make it easy to use.

Reply:
I think we can enforce it.
Have method 
```
public int getDimension(T vector) {} 
```
Then, the caller should provide vector to know about dimension.

Reply:
This is an interesting idea. Let me do this.

Reply:
@heemin32 the problem with this approach is I have to write the implementation of get dimension in each that extends KNNVectorValues. Because T.length is not supported. Do you know if there a way to find the length of T

Reply:
I don't think there is a way to do that. I would like to say this is why generic is not used properly here though not sure if you agree on it :)

Reply:
> I don't think there is a way to do that. I would like to say this is why generic is not used properly here

See this cannot be the reason, because if that is the case then T as generic will never be used then. See this problem of dimension doesn't come if we are using FloatVector Values or ByteVectorValues. So the problem is not with generics, its just we are expecting the dimension before even reading a vector. Its like calling a `docId()` on the Iterator without calling `nextDoc()`. I just added the comment in java doc to ensure that caution is taken.

Reply:
If we use FloatVectorValue and ByteVectorValues, the method will be look like `public int getDimension(float[] vector) {}` `public int getDimension(byte[] vector) {}` each.

Reply:
Or maybe, it is happening because float[] and byte[] is primitive array, and not List<Float> and List<Byte>.

Reply:
IMO, it is fine to have 
```
public abstract int getDimensions(T vector) throws IOException;
```
and let child class to implement it.

Reply:
@heemin32 on thinking more on this, its better to align the interface with `values` interface of Lucene. Hence I am keeping the dimension interface free from `vector` param, but I am not shutting door on the idea of sending vector to get the dimension. 

So for this PR I am adding some validation upfront, which will ensure improper use of interface will throw errors. In the next PR, let me take a stab at completely getting rid of these validations by implementing the dimension logic for BDV. 


Resolving this comment. feel free to un-resolve the comment if you think otherwise. :)